Sources/NIOTLS/NIOTypedApplicationProtocolNegotiationHandler.swift

```diff
@@ -0,0 +1,141 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the SwiftNIO open source project
+//
+// Copyright (c) 2023 Apple Inc. and the SwiftNIO project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftNIO project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+@_spi(AsyncChannel) import NIOCore
+import DequeModule
+
+
+@_spi(AsyncChannel)
+public final class NIOTypedApplicationProtocolNegotiationHandler<NegotiationResult>: ChannelInboundHandler, RemovableChannelHandler, NIOProtocolNegotiationHandler {
```

Comment:
This has loads of common implementation with the existing ALPN handler. Presumably we can factor much of that out to common code.

Reply:
Just pushed a change to create a shared state machine between the two handlers. They still have a bit of implementation in them but this shares the bulk of the underlying logic now.